react-renderer/src/inspector/views/mod/LabeledInput.tsx

```diff
@@ -0,0 +1,258 @@
+        const r = parseInt(hex.slice(1, 3), 16) / 255,
+            g = parseInt(hex.slice(3, 5), 16) / 255,
+            b = parseInt(hex.slice(5, 7), 16) / 255;
+        return ([r, g, b, this.props.eValue.contents[3]] as [number, number, number, number])
+        
+        // this will destroy alpha specified in style.
+        // todo - pass over alpha from style specification? either way color picker prevents alpha styling
+        // maybe there is a custom color picker somewhere that uses alpha... 
+    }
+    // todo - maybe make toCanvas definable in JSON? range doesn't cover all possible use cases for toCanvas
+    public makeRange = () => {  
+        const {inputProps, eAttr} = this.props;
+        return (<input id={eAttr} type= "range" onChange={this.handleChange} min= {toCanvas(inputProps.min!)} max={toCanvas(inputProps.max!)
+        } value={_.round(this.props.eValue.contents)} />)
+    }
+    // for use in makePointRange or potentially other multi-input attributes (would need to be modified)
+    public makeSublabel = (id: string) => {
+        const itemprops = id.split("_");
+        const subindex = itemprops[4];
+        const ptindex = itemprops[3] === "x" ? 0 : 1;
+        console.log(this.state.eValue.contents[+itemprops[1]].contents[+subindex].contents[ptindex]);
+        if (this.props.inputProps.showValue) {
+            return (<label htmlFor={id}>
+            <span>{this.state.eValue.contents[+itemprops[1]].contents[+subindex].contents[ptindex].toString()}</span>
+            {itemprops[0] + itemprops[1] + itemprops[3] + itemprops[4]}</label>) // e.g. "S1y1"
+        }
+        else return (<label htmlFor={id}> {itemprops[0] + itemprops[1] + itemprops[3] + itemprops[4]}</label>); 
+    }
+    public makePointRange = () => {
+        const {inputProps, eAttr} = this.props;
+        const subpaths = this.props.eValue.contents;
+        // todo - refactor the whole file so you can call makerange() and makelabel() with params
+        return (
+        <React.Fragment>
+            {subpaths.map((subpath: SubPath<number>, index: number) => {
+                const ptarray = subpath.contents;
+                // note - prob will crash on bezier stuff
+                return (
+                <React.Fragment key={"S" + index}>
+                {ptarray.map((pt: Elem<number>, subindex: number)=> {
+                    if (pt.tag !== "Pt") throw new Error('No current support for Bezier curves!');
+                    const xid = ["S", index.toString(), eAttr, "x", subindex.toString()].join("_");
+                    const yid = ["S", index.toString(), eAttr, "y", subindex.toString()].join("_");
+                    return (
+                        <React.Fragment key={"S" + index + "pt" + subindex}>
+                        <div style={{display: "inline-block"}} className="sublabinput">
+                            <input id={xid} type="range" onChange={this.handleSubChange} min={toCanvas(inputProps.minX!)}
```

Comment:
Using `id`s anywhere in React is unsupported and, in this case, indicative of a bigger misconception: if you want to reuse an event handler, there are lots of ways to pass that data into it!

One way: wrap the `onChange` in a lambda, and you can pass all the necessary data upstream like so: `onChange={(e) => this.handleSubChange(index, eAttr, subindex, e, ...)`. Or curry the `handleSubChange` for slightly more readability (effectively the same thing).

Reply:
I think both, but it's fine for now